### DISCUSSION TITLE
How can I make plots from `embed` codes to have the same quality as generated using in-document codes?

### DISCUSSION THREAD

**Discussion by juliantao** [quarto-dev/quarto-cli#9486] [Q&A]
### Description

In a manuscript project, I tried to generate a plot in the `index.qmd` from two ways:

1. directly include a code chunk in the `index.qmd` file
2. use `embed` to include the same code chunk from another `qmd` file

I found that the one generated from the `embed`  shortcode has a much lower resolution than the original. I tried to set up `fig-width` or `fig-dpi` in the `yaml` header, it does not seem to have an effect. 

To reproduce:
1. Use the official manuscript template https://github.com/quarto-ext/manuscript-template-vscode.git
2. render the `index.qmd` file, the following plot is generated:
![notebooks-explore-earthquakes-fig-spatial-plot-output-1](https://github.com/quarto-dev/quarto-cli/assets/60116296/14f88cfc-945e-41f6-82c0-9b126b79f3da)

3. In the `index.qmd` file, replace the shortcode `{{< embed notebooks/explore-earthquakes.qmd#fig-spatial-plot >}} with the code directly (note that we need to change the path for the source csv file):
` 
````qmd
```{r}
library(tidyverse)
```


Read a clean version of data:

```{r}
la_palma <- read_csv("./notebooks/la-palma.csv")
```

Create spatial plot:

```{r}
#| label: fig-spatial-plot
#| fig-cap: "Locations of earthquakes on La Palma since 2017"
#| fig-alt: "A scatterplot of earthquake locations plotting latitude
#|   against longitude."
la_palma |> 
  ggplot(aes(Longitude, Latitude)) +
  geom_point(aes(color = Magnitude, size = 40-`Depth(km)`)) +
  scale_color_viridis_c(direction = -1) + 
  scale_size(range = c(0.5, 2), guide = "none") +
  theme_bw()
```
````
4. now render the `index.qmd` again, a new plot is generated.

![fig-spatial-plot-1](https://github.com/quarto-dev/quarto-cli/assets/60116296/11408c66-ed0b-400a-a6a7-1eee62835687)

5. Compare these two plots, the later has a much higher resolution.

**Comment by cscheid** [comment]
You currently cannot, at least not easily. You need the original plot to have been rendered in a single resolution, from which you can embed. This is a generally very hard problem to solve, because embedding the plot while changing the resolution would involve detecting the code that caused the plot to be generated (and its dependencies), and re-running the code while changing the specific resolution settings. We're aware of this limitation but there's no clear technical solution for this that we can deliver.

**Reply by juliantao** [reply]
Thank you for your prompt response. I'm concerned that without resolving this issue, the embed feature might lose its effectiveness in manuscript projects, where high resolution is crucial across all rendering formats. Could you clarify how the default size and resolution for embedded plots are configured in the source code? If feasible, might we consider setting a higher default resolution for these plots? Please disregard this suggestion if it's not a viable option.

**Reply by cscheid** [reply]
> Could you clarify how the default size and resolution for embedded plots are configured in the source code?

It depends on how the embedding works. If you're embedding from an existing .ipynb, we simply use the result that exists in the notebook. But if you're embedding from a .qmd file, then I actually am not sure how that works. It's likely the default DPI in our setup cells: I believe you'll get 96 DPI by default, but I think you should get a higher number if you set `fig-dpi` in your .qmd file.

**Reply by ehudkr** [reply]
I came to a similar conclusion last week.
I had a similar desiderata - I wanted to embed vector-based graphics (svg files) instead raster-based ones (png files). 
Unfortunately, I couldn't get that to work when embedding from .qmd files, only when embedding from .ipynb by defining the inline plotting format to be svg in the first cell of the Notebook:
```ipython
%config InlineBackend.figure_format = "svg"
```

**Comment by cderv** [comment]
I wanted to add, that I do think this difference is related to how rendering from a .qmd works for embed feature. 

This embed feature was designed for notebooks initially, and works through Jupyter by getting the cell result in the rendered notebook. 

If you use `.qmd` with `engine: knitr`, the embed feature currently works by doing a rendering to `.ipynb` format. This is not something well support for this engine and indeed some default option will change from a rendering to HTML for example. 

We have some related issue about this like this one (and others linked in it)
- https://github.com/quarto-dev/quarto-cli/issues/8927

Probably if you use `.qmd` with Jupyter, or `.ipynb` directly as notebook to embed from, you will get closer results. 

We know we need to make some improvement to this intermediate ipynb rendering to make consistent default and other improvment, especially when .qmd uses knitr engine.
